File: backend/src/domain/value-objects/password.ts
import * as bcrypt from 'bcrypt';

export class Password {
  private readonly _hash: string;

  private static readonly SALT_ROUNDS = 10;

  private constructor(hash: string) {
    this._hash = hash;
  }

  static async fromPlainText(plainText: string): Promise<Password> {
    if (!plainText || plainText.trim() === '') {
      throw new Error('Password cannot be empty');
    }
    const hash = await bcrypt.hash(plainText, Password.SALT_ROUNDS);
    return new Password(hash);
  }

  static fromHash(hash: string): Password {
    if (!hash || hash.trim() === '') {
      throw new Error('Password hash cannot be empty');
    }
    return new Password(hash);
  }

  async compare(plainText: string): Promise<boolean> {
    return bcrypt.compare(plainText, this._hash);
  }

  get hash(): string {
    return this._hash;
  }
}

File: backend-go/internal/domain/model/password.go
package model

import (
	"strings"

	"golang.org/x/crypto/bcrypt"
)

const bcryptCost = 10

type Password struct {
	hash string
}

// bcrypt cost 10でハッシュ化する。
func NewPasswordFromPlainText(plainText string) (Password, error) {
	if strings.TrimSpace(plainText) == "" {
		return Password{}, NewError("INVALID_PASSWORD", "Password cannot be empty")
	}
	hash, err := bcrypt.GenerateFromPassword([]byte(plainText), bcryptCost)
	if err != nil {
		return Password{}, err
	}
	return Password{hash: string(hash)}, nil
}

func NewPasswordFromHash(hash string) (Password, error) {
	if strings.TrimSpace(hash) == "" {
		return Password{}, NewError("INVALID_PASSWORD", "Password hash cannot be empty")
	}
	return Password{hash: hash}, nil
}

func (p Password) Compare(plainText string) bool {
	return bcrypt.CompareHashAndPassword([]byte(p.hash), []byte(plainText)) == nil
}

func (p Password) Hash() string {
	return p.hash
}
